README.md
# TileBoard

This is a simple yet highly customizable dashboard for Home Assistant. The main goal of this project was to create a simple dashboard with an easy way to edit and add functionality with minimum knowledge of javascript and html.
Should you have any ideas or questions please post them on the home-assistant forum or create an issue on github.

## Links

* [Discussion on Home Assistant Community](https://community.home-assistant.io/t/new-dashboard-for-ha/57173)
* [Demo Video](https://youtu.be/L8JwzWNAPr8)

## Screenshots

![screen](./images/screenshots/default.png)
![screen](./images/screenshots/transparent.png)
![screen](./images/screenshots/homekit.jpg)
![screen](./images/screenshots/octopus_energy.png)

## How to use

There are a couple of alternative ways of running TileBoard. The options are listed in the recommended order. First being the easiest and the most user friendly.

### Run as a Home Assistant addon

By far the easiest option that also provides auto-update through Home Assistant Supervisor.

Installation instructions: refer to https://github.com/resoai/TileBoard-addon/blob/main/README.md#installation

Check the instructions within the addon on how to edit and customize the configuration.

### Run within a standalone server through Docker

Runs a standalone local HTTP server, independently from the Home Assistant itself.

Setup instructions: refer to https://github.com/resoai/TileBoard/blob/master/docker/README.md

### Run within the public Home Assistant www directory

1. Download the latest release zip file (`Tileboard.zip`) from https://github.com/resoai/TileBoard/releases and unpack to a directory of your choice. (Alternatively you can checkout the repo and build the app manually. Check [CONTRIBUTING](./CONTRIBUTING.md) for more info.)
2. In chosen directory rename `config.example.js` to `config.js` and adjust it for your needs
3. Optionally create an empty `styles/custom.css` file. Everything will work without it but there will be a network request error logged in the console which can be annoying to some.
4. Create a directory called `tileboard` inside `www` directory in HA's config path and copy all unpacked files there.
5. TileBoard will be available at `http://HASS_IP:8123/local/tileboard/index.html` and will prompt you for your login credentials after restarting Home Assistant.

#### !!!WARNING!!!

Files served from the `www` folder (`/local/` url), aren't protected by the Home Assistant authentication. Files stored in this folder, if the URL is known, can be accessed by anybody without authentication. Please make sure that your HA instance is not exposed via Internet or at least that long-lived token is not hardcoded in the config.

## Configure

`config.js` will initialize a global CONFIG object. If you want to have several config files, you can create another one next to `config.js` and name it for example `bedroom.js`. The specific config will be available at `http://HASS_IP:8123/local/tileboard/index.html?config=bedroom`.

This config object has the following fields:

```js
var CONFIG = {
   /* customTheme: specify a custom theme for your dashboard
    * Valid options: null, CUSTOM_THEMES.TRANSPARENT, CUSTOM_THEMES.MATERIAL, CUSTOM_THEMES.MOBILE, CUSTOM_THEMES.COMPACT, CUSTOM_THEMES.HOMEKIT, CUSTOM_THEMES.WINPHONE, CUSTOM_THEMES.WIN95 or a custom theme you have created
    * Default: null. Array supported
    */
   customTheme: null,
   /* transition: The transition effect used between Pages
    * Valid options: TRANSITIONS.ANIMATED, TRANSITIONS.ANIMATED_GPU, TRANSITIONS.SIMPLE
    */
   transition: TRANSITIONS.ANIMATED,
   /* tileSize: The default size (in pixels) of a tile */
   tileSize: Number,
   /* tileMargin: The default margin (in pixels) between tiles */
   tileMargin: Number,
   /* entitySize: Enum size of tile's content (SMALL, NORMAL, BIG)*/
   entitySize: ENTITY_SIZES.NORMAL,
   /* groupMarginCss: CSS margin statement to override the default margin for groups */
   groupMarginCss: '20px 40px',
   /* serverUrl: The URL to your HomeAssistant server */
   serverUrl: 'http://hassio.local:8123',
   /* wsUrl: The URL to your HomeAssistant Websocket connection.
    * If HomeAssistant is behind SSL, replace ws:// with wss://
    */
   wsUrl: 'ws://hassio.local:8123/api/websocket',

   /* authToken: Optional Long live token that you can create in your HomeAssistant. You should never add this if HA is exposed via internet.
    */
   authToken: null,
   /* pingConnection: Set to false disable pinging of the websocket connection.
    * Otherwise, a ping will be sent every five seconds, and if a response is not received in 3 seconds,
    * a reconnect will be attempted. If not included in the config file, setting defaults to true.
    */
   pingConnection: true,
   /* locale: locale used for date and number formats - available locales: it, de, es, fr, pt, ru, nl, pl, en-gb, en-us (default).
    * The system uses localization files from the package angular-i18n.
    * You can add your own locale, if it is not included already, by adding the corresponding file from the angular-i18n
    * package into the 'locales' folder (rename the file so that the file name has format locale_code.js).
    */
   locale: 'en-us',
   /* debug: Toggle for extra debugging information.
    * If enabled, will print info about state changes and entities to console.
    */
   debug: Boolean,
   /* timeFormat: 12 for AM/PM marker, 24 for 24 hour time (default) */
   timeFormat: Number,

   /* googleApiKey: Google API key is required if you are using device tracker tiles along with Google Maps.
    * More info here: https://developers.google.com/maps/documentation/maps-static/usage-and-billing
    */
   googleApiKey: null,
   /* A Mapbox token is required if you are using device tracker tiles along with Mapbox.
    * More info here: https://www.mapbox.com/maps/
    */
   mapboxToken: null,
   /* mapboxStyle: Enter a style URL to change the mapbox style for device tracker tiles.
    * The format of the url is: mapbox://styles/username/style-id
    * If no style URL is entered, the style will default to mapbox/streets-v11.
    */
   mapboxStyle: null,
   /* menuPosition: LEFT (default) or BOTTOM */
   menuPosition: MENU_POSITIONS.LEFT,
   /* hideScrollbar: Hiding horizontal scrollbar */
   hideScrollbar: false,

   /* groupsAlign: Align groups HORIZONTALLY (default) VERTICALLY or GRID
    * The GRID alignment allows you to divide the groups on several lines; in this case it is necessary
    * to specify the row index on which you want to place the group. Each row will contain all the groups
    * with the corresponding index for which the maximum width of the screen will be that of the widest row.
    * With this alignment you will be able to use the screen space to the maximum without having
    * to define groups with a very large height. Group 'row' setting is ignored for HORIZONTALLY and VERTICALLY alignments
    * GRID alignment is not available for MOBILE and WINPHONE themes; if selected fallback to HORIZONTALLY alignment */
   groupsAlign: GROUP_ALIGNS.HORIZONTALLY,

   /* Called when connected to API and fetched the state.
    * Can be called multiple times in case of reconnecting.
    */
   onReady: function () {},

   /* pages: A list of page objects. See documentation on Pages below */
   pages: [],
   /* events: A list of events. See documentation on Events below */
   events: [],
   /* screensaver: A digital picture frame with a clock. Appears when
    * the dashboard has been idle
    * https://github.com/resoai/TileBoard/wiki/Screensaver-configuration
    * (optional)
    */
   screensaver: {
     /* timeout: Idle time (in seconds) before the screensaver will show */
     timeout: 300,
     /* slidesTimeout: Amount of time (in seconds) to show each slide */
     slidesTimeout: 10,

     styles: { fontSize: '40px' },

     /* read more in wiki */
     leftBottom: [{ type: SCREENSAVER_ITEMS.DATETIME }],
     /* slides: Array of paths to pictures and, optionally, associated CSS classes. */
     slides: [
       {bg: 'images/bg1.jpeg', classes:['-slide1']},
       {bg: 'images/bg2.png', classes:['-slide2']},
       {bg: 'images/bg3.jpg', classes:['-slide3']}
     ]
   },

   /* header: object of header. Will be applied globally
    * https://github.com/resoai/TileBoard/wiki/Header-configuration
    * (optional)
    */
   header: DEFAULT_HEADER,

   /* footer: Matching functionality of header, but at the bottom instead of the top
    * https://github.com/resoai/TileBoard/wiki/Header-configuration
    * (optional)
    */
   footer: DEFAULT_HEADER,
}
```

### Pages

Page object can have the following fields:

```js
{
  /* title: The page title (not currently used) */
  title: 'Page title',
  /* bg: Link to the background image */
  bg: 'images/bg1.jpeg',
  /* icon: Page icon for the side menu */
  icon: 'mdi-home-outline', // icon of page (for the side menu)
  /* header: object of header for current page
   * https://github.com/resoai/TileBoard/wiki/Header-configuration
   */
  header: DEFAULT_HEADER,
  /* tileSize: Override the global tileSize value for the current page
     (optional)
   */
  tileSize: Number,
  /* groupMarginCss: Override global groupMarginCss for the current page
   * (optional)
   */
  groupMarginCss: '20px 40px',
  /* hidden: hide page (optional)
   * can be boolean or function that return boolean
   */
  hidden: Boolean | Function,
  /* groups: A list of tile groups. See documentation on Tile Groups below */
  groups: [] // list of tile groups
}
```

### Tile Groups

We divide tiles (cells) into groups on every page. Group object can have the following fields:

```js
{
  /* title: Title to display above the group */
  title: 'Group title',
  /* width: Number of tiles horizontally
   * (optional) Can be calculated automatically
   */
  width: 3,
  /* height: Number of tiles vertically
   * (optional) Can be calculated automatically
   */
  height: 4,
  /* row: Used for positioning the group when using GRID alignment. See description in 'groupsAlign'.
   * (optional) default = 0
   */
  row: 0,
  /* groupMarginCss: Override default margin of tiles for the current group
   * (optional)
   */
  groupMarginCss: '20px 40px',
  /* custom function called on initial layout and when the items are added or removed dynamically.
   * Can be used to adjust items size/position if implementing some dynamic layouts, for example.
   * (optional)
   */
  onLayout: function(page, group) {},
  /* hidden: hide group (optional)
   * can be boolean or function that return boolean
   */
  hidden: Boolean | Function,
  /* items: A list of Tile objects. See documentation on Tiles below */
  items: [],
}
```

### Tiles

Tile Object. [Click here for some real-life examples](TILE_EXAMPLES.md)

```js
{
  /* position: The x,y position of the tile inside the group */
  position: [1, 0],
  /* type: The type of a tile. Valid types are listed below */
  type: TYPES.DEVICE_TRACKER,
  /* id: The entity_id of the device from HomeAssistant (e.g. switch.xyz or light.family_room) */
  id: 'device_tracker.google_maps_228',
  // OPTIONAL
  /* title: Title for the entity. It will use the friendly_name from HomeAssistant if not specified
   * (optional)
   */
  title: 'Tile title',
  /* subtitle: A subtitle to display on the tile
   * (optional)
   */
  subtitle: 'Tile subtitle', // subtitle
  /* width: How many tiles wide this tile should be
   * (optional) (default=1)
   */
  width: 2,
  /* height: How many tiles tall this tile should be
   * (optional) (default=1)
   */
  height: 2,
  /* states: Map a state from HomeAssistant to a different value for TileBoard to display.
   * (optional)
   */
  states: {on: 'Enabled', off: 'Disabled'}, // Object example
  states: function (item, entity) {return entity.state}, // Function example
  /* state: Set a custom state for the tile.
   * (optional)
   */
  state: 'Working', // String example
  state: function (item, entity) {return entity.state}, // Function example
  state: false, // Set to false to disable state
  /* icons: Set the icon for a tile
   * You can use any of the material design icons from https://materialdesignicons.com/
   * as long as they have been published in the [latest npm package](https://github.com/templarian/materialdesign-svg).
   * Use an object or function to map states to icons
   * Note: handling differs for Weather tile. See details in TILE_EXAMPLES.md#weather.
   */
  icons: {on: "mdi-volume-high", off: "mdi-volume-off"}, // Object example
  icons: function (item, entity) {return entity.attributes.icon}, // Function example
  /* icon: Set a static icon for a tile
   * You can use any of the material design icons from https://materialdesignicons.com/
   * as long as they have been published in the [latest npm package](https://github.com/templarian/materialdesign-svg).
   * Note: handling differs for Weather tile. See details in TILE_EXAMPLES.md#weather.
   */
  icon: 'mdi-phone',
  /* customHtml: Replace the icon by a custom HTML content
   * If you define this option, simple tiles will not show the icon. They will show your custom HTML instead.
   * You can use an anonymous function here, or a string that will be interpreted as a template
   */
  customHtml: function (item, entity) {return '<b>' + entity.state + '</b>';},
  customHtml: '<b>&group.lights.state</b>',
  /* bg: Link to a background image for the tile
   * @ and & prefixes are explained below
   */
  bg: '@attributes.entity_picture',
  /* bgSuffix: Same as bg, but with the serverUrl included */
  bgSuffix: '@attributes.entity_picture',
  /* bgOpacity: A decimal between 0 and 1 for the background opacity */
  bgOpacity: 0.5,
  /* theme: Override default theme for the tile */
  theme: TYPES.SWITCH,
  /* slides: A list of slide images to use for the background
   * Currently a maximum of 3 slides are supported
   * (optional)
   */
  slides: [{}, {bg: 'images/slide.jpg'}],

  /* action: Define a custom action on click
   * You can override the default action for any tile type.
   * This function will be evaluated, when the user clicks the tile.
   * Set this to false to disable the default action.
   */
  action: function(item, entity) {return this.$scope.openPopupIframe(item, entity);}
  action: false,
  /* secondaryAction: Define a custom secondary action (on long press)
   * You can override the default secondary action for any tile type.
   * This function will be evaluated, when the user long-presses the tile.
   * Set this to false to disable the default secondary action.
   */
  secondaryAction: function(item, entity) {return this.$scope.openPopupIframe(item, entity);}
  secondaryAction: false,

  /* hidden: hide tile (optional)
   * can be boolean or function that return boolean
   */
  hidden: Boolean | Function,
  /*** TILE SPECIFIC SETTINGS ***/
  /** type: SCRIPT **/
  /* variables: Provides extra variables to the script
   * Can be a function which will be evaluated when script is called.
   */
  variables: {},
  variables: function(item, entity) {return { dynamicVar: true };},
  /** type: SENSOR **/
  /* value: Override sensor value */
  value: '&sensor.bathroom_temp.state',
  /* unit: Override default unit of measurement */
  unit: 'kWh',
  /* filter: Function for filtering/formatting the entity value
   * Is used by default to format a timestamp as time ago, e.g., for the uptime sensor.
   */
  filter: function (value, item, entity) {return Math.round(value);},
  filter: function (value, item, entity) {return timeAgo(value, true);}, // true = exclude the word 'ago'
  /** type: DEVICE_TRACKER **/
  /* slidesDelay: Delay before slide animation starts (optional) */
  slidesDelay: 2,
  /* map: Map provider for showing position inside tile
   * Valid options: 'google', 'mapbox', 'yandex'
   */
  map: 'google',

  /* zoomLevels: zoom levels of map in slides
   * left only one if you what one slide (e.g. [9])
   * (optional)
   */
  zoomLevels: [9, 13],

  /* hideEntityPicture: hide entity picture if you want only maps */
  hideEntityPicture: false,
  /** type: TEXT_LIST **/
  /* list: List of objects with a title, icon, and value */
  list: [{title: 'Kitchen temp', icon: 'mdi-home', value: '&sensor.kitchen_temp.state'}],
  /** type: MEDIA_PLAYER **/
  /* hideSource: Whether the source selector should be hidden
   * Value options: true, false
   */
  hideSource: false,
  /* textSource: localized/customized text for 'Source'
   * (optional)
   */
  textSource: 'Audio Input',
  /** type: SLIDER **/
  /* filter: Function for filtering/formatting the value */

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [2.10.2](https://github.com/resoai/TileBoard/compare/v2.10.1...v2.10.2) (2022-06-30)


### Bug Fixes

* programmatic openPopup can end up with dead tiles ([#843](https://github.com/resoai/TileBoard/issues/843)) ([1a11c5a](https://github.com/resoai/TileBoard/commit/1a11c5a24bdad48f9ee04aae6eb7bf30de6f27cc))

### [2.10.1](https://github.com/resoai/TileBoard/compare/v2.10.0...v2.10.1) (2022-06-20)


### Bug Fixes

* **webmanifest:** add start_url ([#838](https://github.com/resoai/TileBoard/issues/838)) ([eef5368](https://github.com/resoai/TileBoard/commit/eef536801a4d5a8ba3df54fc97e920b313faf0bb))

## [2.10.0](https://github.com/resoai/TileBoard/compare/v2.9.0...v2.10.0) (2022-06-19)


### Features

* make tileboard installable as web apk ([#822](https://github.com/resoai/TileBoard/issues/822)) ([a1dd248](https://github.com/resoai/TileBoard/commit/a1dd248cea665ab760b254cacbcfb82dfd7a4cf8))

## [2.9.0](https://github.com/resoai/TileBoard/compare/v2.8.0...v2.9.0) (2022-02-03)


### Features

* add optional footer, matching functionality of header ([#813](https://github.com/resoai/TileBoard/issues/813)) ([6979fcd](https://github.com/resoai/TileBoard/commit/6979fcddec08f696ec60b65e6d00aebf4f212c8f))
* update dependency @mdi/font to v6 ([#766](https://github.com/resoai/TileBoard/issues/766)) ([384b49d](https://github.com/resoai/TileBoard/commit/384b49dd946b9f78a97e6a5e2e3b40a3b5d53191))

## [2.8.0](https://github.com/resoai/TileBoard/compare/v2.7.0...v2.8.0) (2021-12-04)


### Features

* allow custom classes on screensaver slides ([#796](https://github.com/resoai/TileBoard/issues/796)) ([514abba](https://github.com/resoai/TileBoard/commit/514abba9f82819d1d6686b18bbb11b8e5df0bb06))

## [2.7.0](https://github.com/resoai/TileBoard/compare/v2.6.5...v2.7.0) (2021-11-22)


### Features

* create docker images for standalone use ([#793](https://github.com/resoai/TileBoard/issues/793)) ([73a854e](https://github.com/resoai/TileBoard/commit/73a854e9d121a22b040ff88a126ba00bd30e0f30))

### [2.6.5](https://github.com/resoai/TileBoard/compare/v2.6.4...v2.6.5) (2021-10-28)


### Bug Fixes

* charts not working ([#783](https://github.com/resoai/TileBoard/issues/783)) ([8a62b98](https://github.com/resoai/TileBoard/commit/8a62b988be5ffc8c12fa086ac3bbc53ad34a202d))
* handle web socket API URL that have extra double slashes ([#782](https://github.com/resoai/TileBoard/issues/782)) ([71d0633](https://github.com/resoai/TileBoard/commit/71d063399a3131c1f5e4ba795192e1df291f5358))

### [2.6.4](https://github.com/resoai/TileBoard/compare/v2.6.3...v2.6.4) (2021-10-26)


### Bug Fixes

* **LIGHT:** better styling for the view triggered on long press ([#780](https://github.com/resoai/TileBoard/issues/780)) ([8193f33](https://github.com/resoai/TileBoard/commit/8193f33d532321edbe449532eff469af30aa903f))

### [2.6.3](https://github.com/resoai/TileBoard/compare/v2.6.2...v2.6.3) (2021-08-06)


### Bug Fixes

* **light:** fix light brightness adjustment buttons after HA update ([#746](https://github.com/resoai/TileBoard/issues/746)) ([dd9510f](https://github.com/resoai/TileBoard/commit/dd9510f1e4c484b8d2b3abcb059d70c7006c7718)), closes [#743](https://github.com/resoai/TileBoard/issues/743)

### [2.6.2](https://github.com/resoai/TileBoard/compare/v2.6.1...v2.6.2) (2021-07-21)

### [2.6.1](https://github.com/resoai/TileBoard/compare/v2.6.0...v2.6.1) (2021-05-23)


### Bug Fixes

* **deps:** update dependency node-sass to v6 ([#716](https://github.com/resoai/TileBoard/issues/716)) ([a239003](https://github.com/resoai/TileBoard/commit/a239003475192a4baed3f3e85588d24d7c281896))
* **HISTORY:** fix tooltip time for some locales ([#724](https://github.com/resoai/TileBoard/issues/724)) ([e2c5639](https://github.com/resoai/TileBoard/commit/e2c5639be74647fab2e0e7d539a76415ea42d81b))
* suppress undefined postfix ([#722](https://github.com/resoai/TileBoard/issues/722)) ([d7132ee](https://github.com/resoai/TileBoard/commit/d7132ee257ccf428b1008fdd62aef7a7314c185f))

## [2.6.0](https://github.com/resoai/TileBoard/compare/v2.5.0...v2.6.0) (2021-05-01)


### Features

* onLayout hook triggered on items being added/remove ([#703](https://github.com/resoai/TileBoard/issues/703)) ([94a1846](https://github.com/resoai/TileBoard/commit/94a184604d6f2659a4a6ed1b82f626b7e6affdd9))


### Bug Fixes

* **CLIMATE:** localized/customized lookup for climate options ([f420183](https://github.com/resoai/TileBoard/commit/f420183b2e372f8dc9bc98a07b3363a6434a9079))
* **LIGHT:** show brightness control only when on ([#707](https://github.com/resoai/TileBoard/issues/707)) ([17a5a63](https://github.com/resoai/TileBoard/commit/17a5a6362077673ab89545394d60b9c8735fe8d9))
* **MEDIA_PLAYER:** localized translation for mediaplayer "Source" label ([af9b961](https://github.com/resoai/TileBoard/commit/af9b9618ea5407c27d42e6e14ff6d3cd29eac1ab))

## [2.5.0](https://github.com/resoai/TileBoard/compare/v2.4.1...v2.5.0) (2021-04-20)


### Features

* **SLIDER:** add vertical slider option and improve design ([#672](https://github.com/resoai/TileBoard/issues/672)) ([e713089](https://github.com/resoai/TileBoard/commit/e7130896519a4ea610140960f86ec94c4c4c30cf))

### [2.4.1](https://github.com/resoai/TileBoard/compare/v2.4.0...v2.4.1) (2021-04-15)


### Bug Fixes

* allow overriding REST API domain for ingress ([5987c16](https://github.com/resoai/TileBoard/commit/5987c16fd776219f24ce24f5226777837488a604))
* some issues in default example that triggered console errors ([f95b918](https://github.com/resoai/TileBoard/commit/f95b9184e657066d073e7a17b5cae1c3a15e6293))

## [2.4.0](https://github.com/resoai/TileBoard/compare/v2.3.1...v2.4.0) (2021-04-11)


### Features


variable.js

/**
 * AUTOMATIC TILE SIZING
 *
 * Calculates tile size automatically for different screensizes.
 */

var groupWidth = 3; // count of tiles horizontally
var tileMargin = 5; // px
var groupMargin = 5; // px
var maxTileSize = 150; // px

var tileSize = getTileSize(groupWidth, tileMargin, groupMargin, maxTileSize);

function getTileSize (groupWidth, tileMargin, groupMargin, maxTileSize) {
   var width = window.innerWidth;
   width -= groupMargin * 2 + tileMargin * (groupWidth - 1);
   return Math.min(maxTileSize, +(width / groupWidth).toFixed(1));
}



/**
 * AUTOMATIC CONFIGURATION LOADING
 * 
 * The following snippet updates the configuration with an automatically
 * generated one, as soon as the Api is available.
 */

// When the body has finished loading
angular.element(document.body).ready(function (){
   // Wait also for the Api to be ready
   Api.onReady(function() {
      // Read out the states from HA and do the magic
      Api.getStates(function(res) {
         // The states from HA need some reordering to be useful
         var states = {};
         res.result.forEach(function (state) {
            states[state.entity_id] = state;
         });
         // The $scope variables is where TileBoard loads its views from
         var $scope = angular.element(document.body).scope();
         // Import a group like this:
         $scope.pages[0].groups.unshift(importGroup({
            id: 'group.all_scripts',
            width: 3
            }, states));
         // Import a complete view as a page like this:
         $scope.pages.push(importPage({
            id: 'group.default_view'
            }, states));
         // Finally call the view:updated trigger to, well, update the view
         angular.element(window).triggerHandler('view:updated'); 
      });
   });
});

var CONFIG = {
   customTheme: null, // CUSTOM_THEMES.TRANSPARENT, CUSTOM_THEMES.MATERIAL, CUSTOM_THEMES.MOBILE, CUSTOM_THEMES.COMPACT, CUSTOM_THEMES.HOMEKIT, CUSTOM_THEMES.WINPHONE, CUSTOM_THEMES.WIN95
   transition: TRANSITIONS.ANIMATED_GPU, //ANIMATED or SIMPLE (better perfomance)
   entitySize: ENTITY_SIZES.NORMAL, //SMALL, BIG are available
   tileSize: 150,
   tileMargin: 6,
   serverUrl: "https://"+window.location.hostname+":8123",
   wsUrl: "wss://"+window.location.hostname+":8123/api/websocket",
   authToken: null, // optional long-lived token (CAUTION: only if TileBoard is not exposed to the internet)
   //googleApiKey: "XXXXXXXXXX", // Required if you are using Google Maps for device tracker
   //mapboxToken: "XXXXXXXXXX", // Required if you are using Mapbox for device tracker
   debug: false, // Prints entities and state change info to the console.
   pingConnection: true, //ping connection to prevent silent disconnections

   // next fields are optional
   events: [],
   timeFormat: 24,
   menuPosition: MENU_POSITIONS.LEFT, // or BOTTOM
   hideScrollbar: false, // horizontal scrollbar
   groupsAlign: GROUP_ALIGNS.HORIZONTALLY, // or VERTICALLY

   header: { // https://github.com/resoai/TileBoard/wiki/Header-configuration
      styles: {
         padding: '30px 130px 0',
         fontSize: '28px'
      },
      right: [],
      left: [
         {
            type: HEADER_ITEMS.DATETIME,
            dateFormat: 'EEEE, LLLL dd', //https://docs.angularjs.org/api/ng/filter/date
         }
      ]
   },

   /*screensaver: {// optional. https://github.com/resoai/TileBoard/wiki/Screensaver-configuration
      timeout: 300, // after 5 mins of inactive
      slidesTimeout: 10, // 10s for one slide
      styles: { fontSize: '40px' },
      leftBottom: [{ type: SCREENSAVER_ITEMS.DATETIME }], // put datetime to the left-bottom of screensaver
      slides: [
         { bg: 'images/bg1.jpeg' },
         {
            bg: 'images/bg2.png',
            rightTop: [ // put text to the 2nd slide
               {
                  type: SCREENSAVER_ITEMS.CUSTOM_HTML,
                  html: 'Welcome to the <b>TileBoard</b>',
                  styles: { fontSize: '40px' }
               }
            ]
         },
         { bg: 'images/bg3.jpg' }
      ]
   },*/

   pages: [
      {
         title: 'Main page',
         bg: 'images/bg1.jpeg',
         icon: 'mdi-home-outline', // home icon
         groups: [
            {
               title: 'First group',
               width: 2,
               height: 3,
               items: [
                  {
                     position: [0, 0],
                     width: 2,
                     type: TYPES.TEXT_LIST,
                     id: {}, // using empty object for an unknown id
                     state: false, // disable state element
                     list: [
                        {
                           title: 'Sun.sun state',
                           icon: 'mdi-weather-sunny',
                           value: '&sun.sun.state'
                        },
                        {
                           title: 'Custom',
                           icon: 'mdi-clock-outline',
                           value: 'value'
                        }
                     ]
                  },
                  {
                     position: [0, 1], // [x, y]
                     width: 1,
                     type: TYPES.SENSOR,
                     id: 'updater.updater',
                     state: '@attributes.release_notes' // https://github.com/resoai/TileBoard/wiki/Templates
                  }
               ]
            },

            {
               title: 'Second group',
               width: 2,
               height: 3,
               items: [
                  {
                     position: [0, 0],
                     width: 1,
                     type: TYPES.SLIDER,
                     //id: "input_number.volume",
                     id: {state: 50}, // replace it with real string id
                     state: false,
                     title: 'Custom slider',
                     subtitle: 'Example of subtitle',
                     slider: {
                        min: 0,
                        max: 100,
                        step: 2,
                        request: {
                           type: "call_service",
                           domain: "input_number",
                           service: "set_value",
                           field: "value"
                        }
                     }
                  },
                  {
                     position: [1, 0],
                     width: 1,
                     type: TYPES.SWITCH,
                     //id: "switch.lights",
                     id: {state: 'off'}, // replace it with real string id (e.g. "switch.lights")
                     state: false,
                     title: 'Custom switch',
                     icons: {'off': 'mdi-volume-off', 'on': 'mdi-volume-high'}
                  },
                  {
                     position: [0, 1],
                     type: TYPES.ALARM,
                     //id: "alarm_control_panel.home_alarm",
                     id: { state: 'disarmed' }, // replace it with real string id
                     title: 'Home Alarm',
                     icons: {
                        disarmed: 'mdi-bell-off',
                        pending: 'mdi-bell',
                        armed_home: 'mdi-bell-plus',
                        armed_away: 'mdi-bell',
                        triggered: 'mdi-bell-ring'
                     },
                     states: {
                        disarmed: 'Disarmed',
                        pending: 'Pending',
                        armed_home: 'Armed home',
                        armed_away: 'Armed away',
                        triggered: 'Triggered'
                     }
                  }

               ]
            },

            {
               title: '',
               width: 1,
               height: 3,
               items: [
                  {
                     // please read README.md for more information
                     // this is just an example
                     position: [0, 0],
                     height: 2, // 1 for compact
                     //classes: ['-compact'],
                     type: TYPES.WEATHER,
                     id: {},
                     state: function () {return 'Sunny'}, // https://github.com/resoai/TileBoard/wiki/Anonymous-functions
                     icon: 'clear-day',
                     icons: { 'clear-day': 'clear'},
                     fields: {
                        summary: 'Sunny',
                        temperature: '18',
                        temperatureUnit: 'C',
                        windSpeed: '5',
                        windSpeedUnit: 'kmh',
                        humidity: '50',
                        humidityUnit: '%',
                        list: [
                           'Feels like 16 C'
                           /*
                           'Feels like '
                              + '&sensor.dark_sky_apparent_temperature.state'
                              + '&sensor.dark_sky_apparent_temperature.attributes.unit_of_measurement',
                           'Pressure '
                              + '&sensor.dark_sky_pressure.state'
                              + '&sensor.dark_sky_pressure.attributes.unit_of_measurement',
                           '&sensor.dark_sky_precip_probability.state'
                              + '&sensor.dark_sky_precip_probability.attributes.unit_of_measurement'
                              + ' chance of rain'
                           */
                        ]
                     }
                  }

               ]
            },
            autoGroup({ // use autoGroup() to automatically arrange tiles in a group
               title: 'autoGroup(...) and friends',
               items:
               [
                  ['sun.sun',                  , 'sensor.uptime'   ],
                  [         , 'updater.updater',                   ],
                  [
                     TEMPLATES.SNAPCLIENT('media_player.snapcast_client_00_00_00_00_00_00',[]), // use template functions to make your life easier
                     autoTile            ('media_player.snapcast_client_00_00_00_00_00_00',[]), // use autoTile() to select the right template for you 
                  ]
               ]
            }),
         ]
      },
      {
         title: 'Second page',
         bg: 'images/bg2.png',
         icon: 'mdi-numeric-2-box-outline',
         groups: [
            {
               title: '',
               width: 2,
               height: 3,
               items: [
                  {
                     position: [0, 0],
                     width: 2,
                     title: 'Short instruction',
                     type: TYPES.TEXT_LIST,
                     id: {}, // using empty object for an unknown id
                     state: false, // disable state element
                     list: [
                        {
                           title: 'Read',
                           icon: 'mdi-numeric-1-box-outline',
                           value: 'README.md'
                        },
                        {
                           title: 'Ask on forum',
                           icon: 'mdi-numeric-2-box-outline',
                           value: 'home-assistant.io'
                        },
                        {
                           title: 'Open an issue',
                           icon: 'mdi-numeric-3-box-outline',
                           value: 'github.com'
                        }
                     ]
                  },
                  {
                     position: [0, 1.5],
                     width: 1.5,
                     height: 1,
                     title: 'My Gauge Title',
                     subtitle: '',
                     type: TYPES.GAUGE,
                     id: 'sensor.my_sample_sensor', // Assign the sensor you want to display on the gauge
                     value: function(item, entity) {
                        return entity.state;
                     },
                     settings: {
                        size: 200, // Defaults to 50% of either height or width, whichever is smaller
                        type: 'full', // Options are: 'full', 'semi', and 'arch'. Defaults to 'full'
                        min: 0, // Defaults to 0
                        max: 25000, // Defaults to 100
                        cap: 'round', // Options are: 'round', 'butt'. Defaults to 'butt'
                        thick: 8, // Defaults to 6
                        label: 'My Gauge', // Defaults to undefined
                        append: '@attributes.unit_of_measurement', // Defaults to undefined
                        prepend: '$', // Defaults to undefined
                        duration: 1500, // Defaults to 1500ms
                        thresholds: { 0: { color: 'green'}, 80: { color: 'red' } }, // Defaults to undefined
                        labelOnly: false, // Defaults to false
                        foregroundColor: 'rgba(0, 150, 136, 1)', // Defaults to rgba(0, 150, 136, 1)
                        backgroundColor: 'rgba(0, 0, 0, 0.1)', // Defaults to rgba(0, 0, 0, 0.1)
                        fractionSize: 0, // Number of decimal places to round the number to. Defaults to current locale formatting
                     },
                  },
               ]
            },
         ]
      }
   ],
}
